Add delete room admin endpoint (#7613)
matrix-org/synapse: docs/admin_api/rooms.md

and block these rooms.
It is a combination and improvement of "[Shutdown room](shutdown_room.md)" 
and "[Purge room](purge_room.md)" API.

Shuts down a room. Moves all local users and room aliases automatically to a
new room if `new_room_user_id` is set. Otherwise local users only
leave the room without any information.

The new room will be created with the user specified by the `new_room_user_id` parameter
as room administrator and will contain a message explaining what happened. Users invited
to the new room will have power level `-10` by default, and thus be unable to speak.

If `block` is `True` it prevents new joins to the old room.

This API will remove all trace of the old room from your database after removing
all local users.
Depending on the amount of history being purged a call to the API may take
several minutes or longer.

The local server will only have the power to move local user and room aliases to
the new room. Users on other servers will be unaffected.

## Parameters

The following parameters should be set in the URL:

* `room_id` - The ID of the room.

The following JSON body parameters are available:

> * `new_room_user_id` - Optional. A string representing the user ID of the user that will admin
>                        the new room that all users in the old room will be moved to. If not
>                        set the users will not be moved to a new room and only leave the old room
>                        without any information. Defaults to `None`.

Reviewer: ```suggestion
* `new_room_user_id` - Optional. If set, a new room will be created with this user ID
      as the creator and admin, and all users in the old room will be moved into that
      room. If not set, no new room will be created and the users will just be removed
      from the old room.
```

* `room_name` - Optional. A string representing the name of the room that new users will be
                invited to. Defaults to `Content Violation Notification`
* `message` - Optional. A string containing the first message that will be sent as
              `new_room_user_id` in the new room. Ideally this will clearly convey why the
               original room was shut down. Defaults to `Sharing illegal content on this server
               is not permitted and rooms in violation will be blocked.`

> * `block` - Optional. A boolean if `room_id` will be set on blocking list. The room will be
>             blocked for this server and preventing new joins. Defaults to `True`.

Reviewer: ```suggestion
* `block` - Optional. If set to `true`, this room will be added to a blocking list, preventing future attempts to
  join the room. Defaults to `true`.
```

>             blocked for this server and preventing new joins. Defaults to `True`.

Reviewer: wouldn't it make more sense for this to default to false?

> The following fields are possible in the JSON response body:

Reviewer: ```suggestion
The following fields are returned in the JSON response body:
```

* `kicked_users` - An array of users (`user_id`) that were kicked.
* `failed_to_kick_users` - An array of users (`user_id`) that that were not kicked.
* `local_aliases` - An array of strings representing the local aliases that were migrated from
                    the old room to the new.
* `new_room_id` - A string representing the room ID of the new room.

The API is:

```json
DELETE /_synapse/admin/v1/rooms/<room_id>

{
    "new_room_user_id": "@someuser:example.com",
    "room_name": "Content Violation Notification",
    "message": "Bad Room has been shutdown due to content violations on this server. Please review our Terms of Service.",
    "block": true
}
```

To use it, you will need to authenticate by providing an ``access_token`` for a

> server admin: see `README.rst <README.rst>`_.

Author: I have moved it up.

> A response body like the following is returned:

Author: Parameters are at the end now and we do not need a new heading.

* `new_room_user_id` - Optional. If set, a new room will be created with this user ID
      as the creator and admin, and all users in the old room will be moved into that
      room. If not set, no new room will be created and the users will just be removed

>       from the old room. The user ID must be from local server but it does not have to exist.

Reviewer: this is confusingly worded.

```suggestion
      from the old room. The user ID must be on the local server, but does not necessarily
      have to belong to a registered user.
```
